azure-iot-device/azure/iot/device/iothub/transport/mqtt/mqtt_provider.py

```diff
@@ -132,52 +154,108 @@ def connect(self, password):
 
     def reconnect(self, password):
         """
-        This method reconnects the mqtt broker, possibly because of a password (sas) change
-        Connect should have previously been called.
+        Reconnect to the MQTT broker, using username set at instantiation.
+
+        Connect should have previously been called in order to use this function.
+
+        :param str password: The password for reconnecting with the MQTT broker.
         """
         logger.info("reconnecting transport")
         self._mqtt_client.username_pw_set(username=self._username, password=password)
         self._mqtt_client.reconnect()
 
     def disconnect(self):
         """
-        This method disconnects the mqtt provider. This should be called from the upper transport
-        when it wants to disconnect from the mqtt provider.
+        Disconnect from the MQTT broker.
         """
         logger.info("disconnecting transport")
         self._mqtt_client.disconnect()
+        self._mqtt_client.loop_stop()
```

Comment:
>loop_stop [](start = 26, length = 9)

Did you have a reason for adding this back in?  Is this required or did it just seem like a good idea?  
I'm asking because I've seen this hang before (it waits for all IO to flush before returning).  I think I fixed the deadlock that was causing the hang, but I'm not sure what I think about adding this back in.  I do know that the loop runs in a daemon thread, and I do know that it's harmless to call loop_start multiple times.
I think we should maybe leave it in because I think I fixed the deadlock, but we should be aware that it might hang.

Reply:
I didn't add it back in for any reason other than, it seemed missing based on the docs I was looking at, and there were references to it in docs/tests.

I'm open to re-removing it, but I haven't seen any issues. I say let's keep it for now, and be aware of the potential issue.

---
In reply to: [277871470](https://github.com/Azure/azure-iot-sdk-python-preview/pull/82#discussion_r277871470) [](ancestors = 277871470)